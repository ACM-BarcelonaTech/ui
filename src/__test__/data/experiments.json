[
  {
    "processingConfig": {},
    "lastViewed": null,
    "createdDate": "0000-00-00T00:00:00.000Z",
    "meta": {
      "organism": null,
      "type": "10x"
    },
    "sampleIds": [],
    "description": "Analysis description",
    "experimentId": "'testae48e318dab9a1bd0bexperiment'",
    "projectId": "mock594b-mock-test-ba89-c2c812d3test",
    "experimentName": "Test Project"
  }
]