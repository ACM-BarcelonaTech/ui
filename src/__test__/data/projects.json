[
  {
    "metadataKeys": [],
    "createdDate": "0000-00-00T00:00:00.000Z",
    "experiments": [
      "testae48e318dab9a1bd0experiment1"
    ],
    "name": "Project without samples",
    "description": "Project without samples",
    "lastAnalyzed": null,
    "lastModified": "0000-00-00T00:00:00.000Z",
    "uuid": "4761594b-mock-test-ba89-c2c812d39bb5",
    "samples": []
  },
  {
    "metadataKeys": [
      "Track_1"
    ],
    "createdDate": "0000-00-00T00:00:00.000Z",
    "experiments": [
      "testae48e318dab9a1bd0experiment2"
    ],
    "name": "Project with samples",
    "description": "Project with samples",
    "lastAnalyzed": "0000-00-00T00:00:00.000Z",
    "lastModified": "0000-00-00T00:00:00.000Z",
    "uuid": "380c53b4-mock-test-8091-4c82336d6d49",
    "samples": [
      "e162b6b8-mock-test-b9b3-af5f32f160a2",
      "32084405-mock-test-930b-338de0d2db79",
      "997b3bba-mock-test-9d18-1f794162a392"
    ]
  }
]