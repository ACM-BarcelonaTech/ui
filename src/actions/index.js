import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS, CELL_SETS_COLOR,
} from './actionType';
import { connectionPromise } from '../components/content-wrapper/ContentWrapper';


// eslint-disable-next-line func-names
const loadCellSets = (experimentId) => function (dispatch, getState) {
  if (getState().cellSets.data) {
    return Promise.resolve();
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      data: json.cellSets,
    }),
  ).catch((e) => console.log('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (newState) => function (dispatch, getState) {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  return dispatch({
    type: UPDATE_CELL_SETS,
    data: newState,
  });
};

const createCluster = (cellSetInfo, clusterName, color) => function (dispatch, getState) {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  return dispatch({
    type: CREATE_CLUSTER,
    data: newCluster,
  });
};

// eslint-disable-next-line func-names
const loadCells = (experimentId, requestBody) => function (dispatch, getState) {
  if (getState().cells.data) {
    return Promise.resolve();
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body: requestBody,
    };

    io.emit('WorkRequest', request);

    console.log('emitted!!!');

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      console.log('response! ');
      return dispatch({
        type: LOAD_CELLS,
        data: embedding,
      });
    });
  });
};

// eslint-disable-next-line func-names
const cellSetsColor = (colorData) => function (dispatch, getState) {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return Promise.resolve();
};

// eslint-disable-next-line import/prefer-default-export
export {
  loadCellSets, updateCellSets, createCluster, loadCells, cellSetsColor,
};
