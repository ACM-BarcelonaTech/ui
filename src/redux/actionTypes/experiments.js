const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Experiment saved
 */
const EXPERIMENTS_SAVED = `${EXPERIMENTS}/saved`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

/**
 * Data download is requested
 */
const EXPERIMENTS_DATA_DOWNLOAD_REQUESTED = `${EXPERIMENTS}/downloadRequested`;

/**
 * Download link is ready
 */
const EXPERIMENTS_DATA_DOWNLOAD_READY = `${EXPERIMENTS}/downloadReady`;

/**
 * Download link is clicked
 */
const EXPERIMENTS_DATA_DOWNLOADED = `${EXPERIMENTS}/downloaded`;

/**
 * There is download error
 */
const EXPERIMENTS_DATA_DOWNLOAD_ERROR = `${EXPERIMENTS}/downloadError`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_DATA_DOWNLOAD_REQUESTED,
  EXPERIMENTS_DATA_DOWNLOAD_READY,
  EXPERIMENTS_DATA_DOWNLOADED,
  EXPERIMENTS_DATA_DOWNLOAD_ERROR,
};
