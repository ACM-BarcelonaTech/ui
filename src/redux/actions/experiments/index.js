import createExperiment from './createExperiment';
import saveExperiment from './saveExperiment';
import updateExperiment from './updateExperiment';
import updateExperimentBackendStatus from './updateExperimentBackendStatus';
import loadExperiments from './loadExperiments';

export {
  createExperiment,
  saveExperiment,
  updateExperiment,
  loadExperiments,
  updateExperimentBackendStatus,
};
