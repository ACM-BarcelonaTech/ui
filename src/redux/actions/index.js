/* eslint-disable no-param-reassign */
import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, PUSH_CELL_SETS, CREATE_CLUSTER, CELL_SETS_COLOR,
  UPDATE_GENE_LIST, LOAD_GENE_LIST, SELECTED_GENES, UPDATE_GENE_EXPRESSION,
  LOAD_CELLS, BUILD_HEATMAP_SPEC, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from './actionType';
import connectionPromise from '../../utils/socketConnection';

const loadCellSets = (experimentId) => (dispatch, getState) => {
  if (getState().cellSets.data) {
    return null;
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      experimentId,
      data: json.cellSets,
    }),
  ).catch((e) => console.error('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (experimentId, newState) => (dispatch, getState) => {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  dispatch({
    type: UPDATE_CELL_SETS,
    experimentId,
    data: newState,
  });

  return dispatch(pushCellSets(experimentId));
};

const pushCellSets = (experimentId) => (dispatch, getState) => {
  // If no loaded data exists for our cell sets and some event would
  // trigger a push, we do not want to overwrite our cell sets with
  // empty data.
  if (!getState()?.cellSets?.data) {
    return null;
  }

  return fetch(
    `${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(
        getState().cellSets.data,
        (k, v) => ((k === 'title') ? undefined : v),
      ),
    },
  ).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: PUSH_CELL_SETS,
      experimentId,
      data: json,
    }),
  ).catch((e) => console.error('Error when trying to push cell sets to API: ', e));
};

const createCluster = (experimentId, cellSetInfo, clusterName, color) => (dispatch) => {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  dispatch({
    type: CREATE_CLUSTER,
    experimentId,
    data: newCluster,
  });

  return dispatch(pushCellSets(experimentId));
};

const cellSetsColor = (colorData) => (dispatch, getState) => {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return null;
};

const loadCells = (experimentId, embeddingType) => (dispatch, getState) => {
  if (getState().cells.data) {
    return null;
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const body = {
      name: 'GetEmbedding',
      type: embeddingType,
    };

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      return dispatch({
        type: LOAD_CELLS,
        experimentId,
        data: embedding,
      });
    });
  });
};

const updateGeneList = (experimentId, tableState) => (dispatch, getState) => {
  if (getState().geneList?.tableState === tableState) {
    return null;
  }
  dispatch({
    type: LOAD_GENE_LIST,
  });
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();
    const { geneList } = getState();

    const orderBy = tableState?.sorter.field
      || geneList.tableState?.sorter.field
      || 'dispersions';

    const orderDirection = tableState?.sorter.order
      || geneList.tableState?.sorter.order
      || 'ascend';

    const currentPage = tableState?.pagination.current
      || geneList.tableState?.pagination.current
      || 1;

    const currentPageSize = tableState?.pagination.pageSize
      || geneList.tableState?.pagination.pageSize
      || 1;

    const body = {
      name: 'ListGenes',
      selectFields: ['gene_names', 'dispersions'],
      orderBy,
      orderDirection: (orderDirection === 'ascend') ? 'ASC' : 'DESC',
      offset: ((currentPage - 1) * currentPageSize),
      limit: currentPageSize,
    };

    if (tableState.geneNamesFilter) {
      body.geneNamesFilter = tableState.geneNamesFilter;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const data = JSON.parse(res.results[0].body);

      const { total, rows } = data;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      if (tableState && tableState.pagination) {
        tableState.pagination.total = total;
      }
      return dispatch({
        type: UPDATE_GENE_LIST,
        experimentId,
        data: {
          rows,
          tableState: tableState || geneList.tableState,
        },
      });
    });
  });
};

const loadDiffExpr = (
  experimentId, comparisonType, firstSelectedCluster, secondSelectedCluster,
) => (dispatch) => {
  dispatch({
    type: LOAD_DIFF_EXPR,
  });
  return connectionPromise().then((io) => {
    const ComparisonTypes = {
      One: 'Versus Rest',
      Two: 'Across Sets',
    };

    const requestUuid = uuidv4();

    const body = {
      name: 'DifferentialExpression',
      maxNum: 100,
      cellSet: firstSelectedCluster.key,
    };
    if (comparisonType === ComparisonTypes.One) {
      body.compareWith = 'rest';
    } else {
      body.compareWith = secondSelectedCluster.key;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      let data = {};

      try {
        data = JSON.parse(res.results[0].body);
      } catch (error) {
        console.error(error);
        data = { rows: [] };
      }

      const { rows } = data;
      const total = rows.length;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      return dispatch({
        type: UPDATE_DIFF_EXPR,
        experimentId,
        data: {
          allData: rows,
          total,
        },
      });
    });
  });
};

const updateSelectedGenes = (genes, selected) => (dispatch, getState) => {
  const { selectedGenes, geneExperessionData } = getState();
  selectedGenes.geneList = selectedGenes.geneList || {};
  let newGenesAdded = false;
  genes.forEach((gene) => {
    if (selected) {
      if (!selectedGenes.geneList[gene]) {
        // This will present a dict of newly selected genes from gene table
        // { <Gene Name>: <Is the gene experession resolved?> }
        selectedGenes.geneList[gene] = false;
        newGenesAdded = true;
      }
    } else {
      delete selectedGenes.geneList[gene];
      const foundGene = geneExperessionData.data.findIndex((g) => g.geneName === gene);
      geneExperessionData.data.splice(foundGene, 1);
    }
  });
  dispatch({
    type: SELECTED_GENES,
    data: { newGenesAdded },
  });
  dispatch({
    type: UPDATE_HEATMAP_SPEC,
    data: {
      genes: geneExperessionData.data,
      rendering: true,
    },
  });
  setTimeout(() => {
    // Running this action with a delay to allow rerendering of heatmap
    dispatch({
      type: UPDATE_HEATMAP_SPEC,
      data: { rendering: false },
    });
  }, 50);
};

const loadGeneExpression = (experimentId) => (dispatch, getState) => {
  const { geneExperessionData } = getState();
  if (geneExperessionData?.isLoading) {
    return null;
  }
  const { newGenesAdded, geneList } = getState().selectedGenes;
  const newGeneBatch = [];
  if (newGenesAdded) {
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(geneList)) {
      if (!value) {
        newGeneBatch.push(key);
        geneList[key] = true;
      }
    }
  }
  dispatch({
    type: SELECTED_GENES,
    data: {
      newGenesAdded: false,
    },
  });
  if (newGeneBatch.length) {
    dispatch({
      type: UPDATE_GENE_EXPRESSION,
      data: {
        isLoading: true,
      },
    });
    return connectionPromise().then((io) => {
      const requestUuid = uuidv4();

      const body = {
        name: 'GeneExpression',
        cellSets: 'all',
        genes: newGeneBatch,
      };

      const request = {
        uuid: requestUuid,
        socketId: io.id,
        experimentId,
        timeout: '2021-01-01T00:00:00Z',
        body,
      };

      io.emit('WorkRequest', request);

      io.on(`WorkResponse-${requestUuid}`, (res) => {
        const heatMapData = JSON.parse(res.results[0].body);
        const { data } = getState().geneExperessionData;
        if (data) {
          Array.prototype.push.apply(heatMapData.data, data);
        }
        dispatch({
          type: UPDATE_GENE_EXPRESSION,
          data: {
            heatMapData,
            isLoading: false,
          },
        });
        return dispatch({
          type: BUILD_HEATMAP_SPEC,
          data: {
            geneExperessionData: heatMapData,
          },
        });
      });
    });
  }
};

export {
  loadCellSets,
  updateCellSets,
  pushCellSets,
  createCluster,
  loadCells,
  cellSetsColor,
  updateGeneList,
  updateSelectedGenes,
  loadGeneExpression,
  loadDiffExpr,
};
