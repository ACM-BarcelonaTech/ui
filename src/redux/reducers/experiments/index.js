import initialState from './initialState';
import {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_DATA_DOWNLOAD_REQUESTED,
  EXPERIMENTS_DATA_DOWNLOAD_STARTED,
  EXPERIMENTS_DATA_DOWNLOAD_ERROR,
} from '../../actionTypes/experiments';
import experimentsCreate from './experimentsCreate';
import experimentsUpdate from './experimentsUpdate';
import experimentsDelete from './experimentsDelete';
import experimentsLoading from './experimentsLoading';
import experimentsLoaded from './experimentsLoaded';
import experimentsError from './experimentsError';
import experimentsSaving from './experimentsSaving';
import exprimentsSaved from './experimentsSaved';
import experimentsDownloadRequested from './experimentsDownloadRequested';
import experimentsDownloadStarted from './experimentsDownloadStarted';
import experimentsDownloadError from './experimentsDownloadError';

const experimentsReducer = (state = initialState, action) => {
  switch (action.type) {
    case EXPERIMENTS_CREATED: {
      return experimentsCreate(state, action);
    }

    case EXPERIMENTS_UPDATED: {
      return experimentsUpdate(state, action);
    }

    case EXPERIMENTS_DELETED: {
      return experimentsDelete(state, action);
    }

    case EXPERIMENTS_LOADING: {
      return experimentsLoading(state, action);
    }

    case EXPERIMENTS_LOADED: {
      return experimentsLoaded(state, action);
    }

    case EXPERIMENTS_ERROR: {
      return experimentsError(state, action);
    }

    case EXPERIMENTS_SAVING: {
      return experimentsSaving(state, action);
    }

    case EXPERIMENTS_SAVED: {
      return exprimentsSaved(state, action);
    }

    case EXPERIMENTS_DATA_DOWNLOAD_REQUESTED: {
      return experimentsDownloadRequested(state, action);
    }

    case EXPERIMENTS_DATA_DOWNLOAD_STARTED: {
      return experimentsDownloadStarted(state, action);
    }

    case EXPERIMENTS_DATA_DOWNLOAD_ERROR: {
      return experimentsDownloadError(state, action);
    }
    default: {
      return state;
    }
  }
};

export default experimentsReducer;
