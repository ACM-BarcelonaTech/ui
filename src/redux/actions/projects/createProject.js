import moment from 'moment';
import { v4 as uuidv4 } from 'uuid';

import fetchAPI from '../../../utils/fetchAPI';
import { isServerError, throwIfRequestFailed } from '../../../utils/fetchErrors';

import {
  PROJECTS_ERROR,
  PROJECTS_CREATE,
  PROJECTS_SAVING,
} from '../../actionTypes/projects';

import { projectTemplate } from '../../reducers/projects/initialState';
import createExperiment from '../experiments/createExperiment';
import endUserMessages from '../../../utils/endUserMessages';
import pushNotificationMessage from '../../../utils/pushNotificationMessage';

const createProject = (
  projectName,
  projectDescription,
  newExperimentName,
) => async (dispatch) => {
  const createdDate = moment().toISOString();

  const newProjectUuid = uuidv4();

  // Always create an experiment for a new project
  // required because samples DynamoDB require experimentId
  const newExperiment = await dispatch(createExperiment(newProjectUuid, newExperimentName));

  dispatch({
    type: PROJECTS_SAVING,
    payload: {
      message: endUserMessages.SAVING_PROJECT,
    },
  });

  const newProject = {
    ...projectTemplate,
    name: projectName,
    description: projectDescription,
    uuid: newProjectUuid,
    experiments: [newExperiment.id],
    createdDate,
    lastModified: createdDate,
  };

  try {
    const url = `/v1/projects/${newProjectUuid}`;
    try {
      const response = await fetchAPI(
        url,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(newProject),
        },
      );

      const json = await response.json();

      throwIfRequestFailed(response, json, endUserMessages.ERROR_SAVING);
    } catch (e) {
      let { message } = e;

      if (!isServerError(e)) {
        console.error(`fetch ${url} error ${message}`);
        message = endUserMessages.ERROR_SAVING;
      }

      dispatch({
        type: PROJECTS_ERROR,
        payload: {
          error: message,
        },
      });

      pushNotificationMessage('error', message);
      return Promise.reject(message);
    }

    dispatch({
      type: PROJECTS_CREATE,
      payload: { project: newProject },
    });
  } catch (e) {
    pushNotificationMessage('error', endUserMessages.ERROR_SAVING);
  }
};

export default createProject;
