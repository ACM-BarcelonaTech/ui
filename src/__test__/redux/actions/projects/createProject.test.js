import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 as uuidv4 } from 'uuid';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import createProject from 'redux/actions/projects/createProject';
import initialProjectsState from 'redux/reducers/projects';
import { saveProject } from 'redux/actions/projects';
import { createExperiment } from 'redux/actions/experiments';
import {
  PROJECTS_CREATE, PROJECTS_SAVING, PROJECTS_SAVED, PROJECTS_ERROR,
} from 'redux/actionTypes/projects';

// import endUserMessages from 'utils/endUserMessages';

const mockStore = configureStore([thunk]);

jest.mock('uuid');
const mockedProjectUuid = 'random-project-uuid';
uuidv4.mockImplementation(() => mockedProjectUuid);

jest.mock('../../../../redux/actions/projects/saveProject');
saveProject.mockImplementation(() => async () => { });

jest.mock('../../../../redux/actions/experiments/createExperiment');
createExperiment.mockImplementation((projectUuid, projectName) => async () => ({
  name: projectName,
  uuid: projectUuid,
}));

jest.mock('../../../../utils/pushNotificationMessage');
pushNotificationMessage.mockImplementation(() => async () => { });

enableFetchMocks();

describe('createProject action', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.doMock();

    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  const mockProjectName = 'test project';
  const mockProjectDescription = 'test project description';
  const mockExperimentName = 'mockExperimentName';

  const mockFetchErrorMessage = 'someFetchError';

  it('Works correctly when there are no errors', async () => {
    const store = mockStore({
      projects: initialProjectsState,
    });

    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    await store.dispatch(
      createProject(mockProjectName, mockProjectDescription, mockExperimentName),
    );

    expect(createExperiment).toHaveBeenCalledWith(mockedProjectUuid, mockExperimentName);

    const actions = store.getActions();

    expect(actions[0].type).toEqual(PROJECTS_SAVING);
    expect(actions[1].type).toEqual(PROJECTS_SAVED);
    expect(actions[2].type).toEqual(PROJECTS_CREATE);

    // Created project is correct
    expect(actions[2].payload).toMatchSnapshot();
  });

  it('Shows error message when there was a fetch error', async () => {
    const store = mockStore({
      projects: initialProjectsState,
    });

    fetchMock.mockResponse(JSON.stringify({ message: mockFetchErrorMessage }), { url: 'mockedUrl', status: 400 });

    await expect(
      store.dispatch(
        createProject(mockProjectName, mockProjectDescription, mockExperimentName),
      ),
    ).rejects.toEqual(mockFetchErrorMessage);

    const actions = store.getActions();

    expect(actions[0].type).toEqual(PROJECTS_SAVING);

    expect(actions[1].type).toEqual(PROJECTS_ERROR);

    expect(pushNotificationMessage).toHaveBeenCalled();
  });
});
