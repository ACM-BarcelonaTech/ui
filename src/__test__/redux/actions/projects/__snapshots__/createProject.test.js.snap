// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProject action Works correctly when there are no errors 1`] = `
Object {
  "createdDate": "2020-01-01T00:00:00.000Z",
  "description": "test project description",
  "experiments": Array [
    null,
  ],
  "lastAnalyzed": null,
  "lastModified": "2020-01-01T00:00:00.000Z",
  "metadataKeys": Array [],
  "name": "test project",
  "samples": Array [],
  "uuid": "random-project-uuid",
}
`;

exports[`createProject action Works correctly when there are no errors 2`] = `
Object {
  "project": Object {
    "createdDate": "2020-01-01T00:00:00.000Z",
    "description": "test project description",
    "experiments": Array [
      undefined,
    ],
    "lastAnalyzed": null,
    "lastModified": "2020-01-01T00:00:00.000Z",
    "metadataKeys": Array [],
    "name": "test project",
    "samples": Array [],
    "uuid": "random-project-uuid",
  },
}
`;
